### PR TITLE
(BKR-1432) Test if puppet module is directly in source host root

### PR DESCRIPTION
**Pull request by derdanne**
See https://tickets.puppetlabs.com/browse/BKR-1432

### CODE REVIEW

**Comment by pinkypie** [comment]
@puppetlabs-jenkins retest this please

**Comment by mchllweeks** [comment]
@puppetlabs-jenkins restest this please

**Comment by kevpl** [comment]
@derdanne this looks like a good fix so far! In order to get this ready for merging, we'll need these tasks to be done for this PR:

- add a spec test so that we don't break this functionality in the future
- pre-pend your ticket ref to your git commit message
- add a changelog entry to the CHANGELOG file so this will be reported with the release that it's in

Once those are done, we'll look at pushing this through our automation & towards release. Thanks for the fix!

**Comment by mchllweeks** [comment]
Last comment still current.  If no motion is made on this PR in another week, we'll close this PR due to inactivity.

**Comment by mchllweeks** [comment]
Closed due lack of activity.

**Comment by jbernardini33** [comment]
@derdanne, we've seen another instance of this issue. Are you interested in completing the tasks requested by @kevpl above to get this merged?

**Comment by derdanne** [comment]
Hi @jbernardini33, i think i can complete the mentioned tasks during next week.

**Comment by derdanne** [comment]
Hi, i've written a test, but since i have no access to the output of the ci, i can't verify this test is working properly. Do you have any documentation how to run the test suite locally?

**Comment by mchllweeks** [comment]
jenkins test this please

**Comment by mchllweeks** [comment]
@derdanne Could you please rebase onto master to get the latest changes?  I believe that is why the acceptance tests are failing.

**Comment by kevpl** [comment]
@derdanne we don't typically take PRs that have merge commits in them. We would prefer that you rebase your branch onto master itself, so it doesn't show up in your changes history.

Another option if this is too painful now that the merge commit exists in this branch is to start a new branch from master & cherry-pick the commit you have here onto that branch. Let me know if you have any questions about this.

**Comment by mchllweeks** [comment]
jenkins test this please

File: spec/beaker/host_spec.rb [hunk 1]
```diff
@@ -487,6 +487,38 @@ module Beaker
...
+          create_files( @fileset1 )
+          create_files( @fileset2 )
+          create_files( @fileset3 )
+        end
+
+        it "should create target dirs with correct path seperator" do
+          created_target_path = File.join(target_path, File.basename(source_path))
+          expect( host ).to receive( :mkdir_p ).with("#{created_target_path}/tests")
+          expect( host ).to receive( :mkdir_p ).with("#{created_target_path}/tests/tests2")
+          expect( host ).to receive( :mkdir_p ).with("#{created_target_path}/tests/tests3")
```
Comment (mchllweeks) [inline spec/beaker/host_spec.rb]: @derdanne I believe you are missing the call 'host.do_scp_to *args' for this test.

**Comment by mchllweeks** [comment]
@derdanne Tests are failling with the below error.  I commented on the file what I think needs to be done to fix the test.  

13:41:41   1) Beaker::Host do_scp_to using an ignore array with an absolute source path in host root should create target dirs with correct path seperator
13:41:41      Failure/Error: expect( host ).to receive( :mkdir_p ).with("#{created_target_path}/tests")

**Comment by mchllweeks** [comment]
retest this please

File: spec/beaker/host_spec.rb [hunk 2]
```diff
@@ -489,6 +489,41 @@ module Beaker
...
+
+          host.do_scp_to *args
+        end
```
Comment (mchllweeks) [inline spec/beaker/host_spec.rb]: @derdanne The variable *args is not set in your test.

Comment (derdanne) [reply spec/beaker/host_spec.rb]: @mchllweeks Yes - didn't see that it is not globally defined, fixed it!

**Comment by derdanne** [comment]
retest this please

**Comment by mchllweeks** [comment]
retest this please

**Comment by mchllweeks** [comment]
@derdanne Tests are still failing.  Here is the error:

08:48:15 Failures:
08:48:15 
08:48:15   1) Beaker::Host do_scp_to using an ignore array with an absolute source path in host root should create target dirs with correct path seperator
08:48:15      Failure/Error: if self['ip'] && (@connection.ip != self['ip'])
08:48:15        #<Double :connection> received unexpected message :ip with (no args)
08:48:15      # ./lib/beaker/host.rb:281:in `connection'
08:48:15      # ./lib/beaker/host.rb:439:in `do_scp_to'
08:48:15      # ./spec/beaker/host_spec.rb:531:in `block (4 levels) in <module:Beaker>'
08:48:15 
08:48:15 Finished in 5.2 seconds (files took 0.97913 seconds to load)
08:48:15 1389 examples, 1 failure
08:48:15 
08:48:15 Failed examples:
08:48:15 
08:48:15 rspec ./spec/beaker/host_spec.rb:517 # Beaker::Host do_scp_to using an ignore array with an absolute source path in host root should create target dirs with correct path seperator

**Comment by derdanne** [comment]
@mchllweeks OK, can you give it another try?

**Comment by mchllweeks** [comment]
retest this please

**Comment by kevpl** [comment]
@derdanne we're still seeing spec failures running these tests. You can reproduce the exact failure we're seeing locally by running this command in your beaker repo:
```console
$ bundle exec rspec spec/beaker/host_spec.rb

Beaker::Host
  creates a windows host given a windows config
  defaults to a unix host
  can be read like a hash
...
Failures:

  1) Beaker::Host do_scp_to using an ignore array with an absolute source path in host root should create target dirs with correct path seperator
     Failure/Error: result = connection.scp_to(source_file, target, options)
       #<Double :connection> received unexpected message :scp_to with ("/source", "target", {:dry_run=>false})
     # ./lib/beaker/host.rb:439:in `do_scp_to'
     # ./spec/beaker/host_spec.rb:535:in `block (4 levels) in <module:Beaker>'

Finished in 0.48968 seconds (files took 0.8627 seconds to load)
74 examples, 1 failure
```
Once you get these passing locally, you can ensure we get past the spec stage entirely by running specs against the whole repo by removing the file path from the command above.

**Comment by derdanne** [comment]
@kevpl ok, thanks, that was the missing hint - i could test it now locally, and spec tests are passing now!

@mchllweeks Please run the tests again (hopefully last time on this PR ;) ).

**Comment by derdanne** [comment]
@mchllweeks Hey, can you please check, why the testsuite failed? All rspec tests have passed locally.

**Comment by mchllweeks** [comment]
@derdanne These failures are unrelated to your changes.  Once we get these passing again, I will re-run the tests for this pr so we can merge your change.

**Comment by kevpl** [comment]
retest this please

**Pull request merged**